packages/app/src/cli/services/generate/shop-import/declarative-definitions.ts

```diff
@@ -0,0 +1,742 @@
+
+        metaobjectCount++
+        break
+      }
+      case 'not_app_reserved': {
+        break
+      }
+    }
+  }
+
+  for (const {items, ownerType, graphQLOwner} of metafieldNodes) {
+    for (const node of items) {
+      const result = convertMetafield(node, ownerType)
+      switch (result.status) {
+        case 'ok': {
+          const {metafield, namespace, key, patches: metafieldPatches} = result
+
+          if (!toDeclare[ownerType]) {
+            toDeclare[ownerType] = {
+              metafields: {},
+            }
+          }
+
+          const dcddOwnerMetafields = toDeclare[ownerType].metafields
+          const namespaceMetafields = dcddOwnerMetafields[namespace] ?? {}
+          dcddOwnerMetafields[namespace] = namespaceMetafields
+
+          namespaceMetafields[key] = metafield
+          tomls.push(
+            patchesToToml(
+              metafieldPatches,
+              `namespace: ${
+                namespace === 'app' ? '$app' : `$app:${namespace}`
+              } key: ${key} owner_type: ${graphQLOwner}`,
+            ),
+          )
+
+          metafieldCount++
+          break
+        }
+        case 'not_app_reserved': {
+          break
+        }
+      }
+    }
+  }
+
+  const tomlContent = tomls.join('\n')
+
+  return {
+    tomlContent,
+    metafieldCount,
+    metaobjectCount,
+  }
+}
+
+async function _importDeclarativeDefinitions(options: ImportDeclarativeDefinitionsOptions) {
+  const adminSession = await renderSingleTask({
+    title: outputContent`Connecting to shop`,
+    task: async () => {
+      return createAdminApiSessionForShop(options)
+    },
+  })
+  const shopName = adminSession.storeFqdn
+
+  const dcddOwnerToGraphQLMapping: {
+    [key in MetafieldOwners]?: MetafieldOwnerType
+  } = {
+    article: 'ARTICLE',
+    blog: 'BLOG',
+    collection: 'COLLECTION',
+    company: 'COMPANY',
+    company_location: 'COMPANY_LOCATION',
+    location: 'LOCATION',
+    market: 'MARKET',
+    order: 'ORDER',
+    page: 'PAGE',
+    product: 'PRODUCT',
+    customer: 'CUSTOMER',
+    delivery_customization: 'DELIVERY_CUSTOMIZATION',
+    delivery_method: 'DELIVERY_METHOD',
+    delivery_option_generator: 'DELIVERY_OPTION_GENERATOR',
+    discount: 'DISCOUNT',
+    draft_order: 'DRAFTORDER',
+    fulfillment_constraint_rule: 'FULFILLMENT_CONSTRAINT_RULE',
+    gift_card_transaction: 'GIFT_CARD_TRANSACTION',
+    order_routing_location_rule: 'ORDER_ROUTING_LOCATION_RULE',
+    payment_customization: 'PAYMENT_CUSTOMIZATION',
+    selling_plan: 'SELLING_PLAN',
+    shop: 'SHOP',
+    validation: 'VALIDATION',
+    variant: 'PRODUCTVARIANT',
+    cart_transform: 'CARTTRANSFORM',
+  }
+
+  const metafieldLoadResults: {
+    metafields: PaginatedQueryResult<MetafieldForImportFragment>
+    ownerType: MetafieldOwners
+    graphQLOwner: MetafieldOwnerType
+  }[] = []
+
+  await renderTasks(
+    Object.entries(dcddOwnerToGraphQLMapping).map(([dcddOwner, graphQLOwner]) => ({
```

Comment:
Instead of loading metafields for all the owner types and dealing with `ACCESS_DENIED` errors, couldn't we check the app scopes first to request only the accessible ones?

Reply:
It's not possible for us to judge these from the client: the only way to know if you have sufficient access is to try to use your scopes.